### ISSUE TITLE
How to represent tangent bundle?

### ISSUE THREAD

**Issue by dlfivefifty** [JuliaApproximation/SphericalHarmonics.jl#5]
When we do vector-spherical harmonics we'll want 
```julia
V = VectorSphericalHarmonics()
eltype(V)
``` 
to be a type representing the tangent bundle.

**Comment by daanhb** [comment]
@dlfivefifty not really. There is no notion of a manifold, let alone tangent spaces.

The only thing I did is make sure that mappings are not assumed to map between spaces of the same dimension, to enable working with a lower-dimensional parameter domain. A `UnitCircle` has a [parameterization](https://github.com/JuliaApproximation/DomainSets.jl/blob/f960152744cf16bbd458ed07c9f9b300728d048e/src/domains/ball.jl#L207) for example, which goes from 1d to 2d, and it has a left inverse which maps back. I guess that could be done for a sphere, and the jacobian of the map could generically be used to verify whether a vector is in the tangent space. But everything on top of that has yet to be invented.

**Comment by dlfivefifty** [comment]
OK. I think the right way to think of this is as a generalisation of a dual number, where the value is a `SphericalCoordinate <: SVector{3}` and the dual part is `TangentVector <: SVector{3}`.

Probably best to think how a circle in the complex plane would work first. One option would be to treat the tangent vector using lazy arrays, e.g., for something in the tangent space with constant `c` at the point `(1/sqrt(2),1/sqrt(2))` would be represented via:
```julia
TangentDual(SVector(1/sqrt(2), 1/sqrt(2)), ApplyVector(*, c, SVector(1/sqrt(2),-1/sqrt(2))))
```
(Makes me wonder if we can have `ApplyArray` conform to the static array interface...)

**Comment by daanhb** [comment]
Looks interesting, just let me know if there is anything you feel belongs to DomainSets. I wasn't thinking of differential geometry myself, but I would like to expand on the parameterization of domains (or a suitable interface for it). In fact that was part of the motivation for improving maps.
